### DISCUSSION TITLE
parameter to update an interval day to month

### DISCUSSION THREAD

**Discussion by rob975** [psycopg/psycopg#909] [Q&A]
I have an `interval day to month` column I want to update using parameters.
When reading the column I get a `timedelta` with days computed as `years * 365 + months * 30`.
When using the same `timedelta` as a parameter to update the column, it is updated to `0`.
Everything works fine with `interval day to second` columns.

psycopg version is 3.2.1 and server version is 16.0.4. Below are sample code and output.

Am I missing something?

## sql code
```sql
create table test_interval (
  test_id  smallserial,
  test_iym interval year to month,
  test_ids interval day to second(0),
  primary key (test_id));

insert into test_interval (test_iym, test_ids) values
(interval '+123-4' year to month, interval '+123 04:05:06' day to second(0));

select test_iym, test_ids from test_interval;
```

## python code
```python
from os import environ
import psycopg

print(f"{psycopg.__version__=}")
conn = psycopg.connect(conninfo=environ["DB_CONN"])
print(f"{conn.info.server_version=}")
cur = conn.cursor()

# sql statements
_select = "select test_iym, test_ids from test_interval"
_update = "update test_interval set test_iym = %(test_iym)s, test_ids = %(test_ids)s"

# read values from db
cur.execute(_select)
values = cur.fetchone()
print(values)

# update with values previously read
params = dict(zip(("test_iym", "test_ids"), values))
cur.execute(_update, params)

# read new values
cur.execute(_select)
values = cur.fetchone()
print(values)
```

## output
```console
psycopg.__version__='3.2.1'
conn.info.server_version=160004
(datetime.timedelta(days=45015), datetime.timedelta(days=123, seconds=14706))
(datetime.timedelta(0), datetime.timedelta(days=123, seconds=14706))
```

**Comment by dvarrazzo** [comment]
This boils down to Postgres interpreting:

```
piro=# select interval '45015 days 0:00:00' year to month;
┌──────────┐
│ interval │
├──────────┤
│ 00:00:00 │
└──────────┘
(1 row)
```

Now, I use Postgres since 7.3, circa 2002, I have pored over its manuals, probably even printed them at some point in the past (two pages per sheet, both sides) - for better poring - and I have NEVER seen this feature. I assume it is part of the SQL standard because sometimes between 1965 and 1970 it might have saved some byte (definition of byte probably not even standardised on being 8 bits).

Anyways, Postgres is apparently able to tell you how many days are an interval year-to-month but it's clearly not able to say how many year-to-months are in a number of days. I don't think you should be using this half-baked feature. It performs a quiet truncation (the documentation subtly says that these keywords _restrict the set of stored fields_: it doesn't say that they validate the input, they just plonk some data. I think you would be better off to put a constraint on this field or to write a custom domain.

On the Psycopg side, we do make an approximation, because Python's datetime.timedelta doesn't have separate fields for years and months so we can only adapt it to a number of days. The conversion is not perfect, but, better than this, we should have only thrown an error if intervals contain years or months, and I believe it would have made more people unhappy. Because, at the time of running a query, Psycopg doesn't know the field it will write to, it isn't  able to infer "this is a restricted interval, therefore I need to convert the days to months and years".

For extra lols, Python and Postgres disagree on how many days are in 123 years and 4 months:

```
piro=# select extract (epoch from interval '+123-4' year to month) / 60 / 60 / 24;
┌────────────────────┐
│      ?column?      │
├────────────────────┤
│ 45045.750000000000 │
└────────────────────┘
```

This means that Postgres calculates a year as being 365.25 days - how Julian of theirs - however it doesn't calculate the leapness of the months.

Another amusing bit before closing this ticket: can you guess what `select interval '+123' year to month` returns? Do you find it reasonable?

To wrap up:

- truncation of the days is performed by Postgres
- because Python doesn't have years and months in a timedelta, it has to make do with days
- converting days back to months and years is an ambiguous approximation
- we cannot perform this approximation by default, because we would corrupt days-based timedeltas
- we cannot infer that we will write in a year-to-month 

therefore I don't think we can do any better.

---

One final note: if you need to keeps the years-and-months semantics, you can [write loader/dumper](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html) for [`dateutil.relativedelta`](https://dateutil.readthedocs.io/en/stable/relativedelta.html). At a glance (I haven't looked into it further), it does seem to behave like Postgres and it should keep the years and months separate, which is fundamentally what you need. We don't have immediate plans to write such loader/dumper ourselves, but feel free to sponsor us a couple of days of work if you want them.

Thank you for the amusing discovery.
